### ISSUE TITLE
graph break when training LoRA

### ISSUE THREAD

**Issue by felipemello1** [pytorch/pytorch#141039]
### 🐛 Describe the bug

In torchtune, we offer LoRA training, which consists of freezing transformer layers and adding just a few matrices that are trainable.

In this setting,  the input to the first transformer doesnt require gradients. However, the intput to the second transformer does. You can see the loop [here](https://github.com/pytorch/torchtune/blob/1814feb17a9038e1a8053ce93111bd26774a8db6/torchtune/modules/transformer.py#L633).

We compile [each transformer block](https://github.com/pytorch/torchtune/blob/1814feb17a9038e1a8053ce93111bd26774a8db6/torchtune/training/_compile.py#L52) independently, to avoid long compilation times. Because of that, i get a graph break: 
```
"0/0: tensor 'L['x']' requires_grad mismatch. expected requires_grad=0"
```

Is there a mark_dynamic equivalent for required_grad?

tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmptwNd6e/index.html
To reproduce: 
```
import torch
import torch.nn as nn


# Define a simple layer class
class SimpleLayer(nn.Module):
    def __init__(self, input_dim):
        super(SimpleLayer, self).__init__()
        # frozen
        self.linear_1 = nn.Linear(input_dim, input_dim)
        for param in self.linear_1.parameters():
            param.requires_grad = False

        # requires grad
        self.linear_2 = nn.Linear(input_dim, input_dim)
        for param in self.linear_2.parameters():
            param.requires_grad = True

    def forward(self, x):
        x = self.linear_1(x)
        x = self.linear_2(x)
        return x


# Create dummy data
batch_size, input_dim = 2, 4

# Input tensor with requires_grad=False
x = torch.randn(batch_size, input_dim, requires_grad=False)

# Initialize layers
layers = [
    SimpleLayer(input_dim),
    SimpleLayer(input_dim),
]

# Compile the layers
compiled_layers = [torch.compile(layer, backend="inductor") for layer in layers]

# Forward pass through each layer
for layer in compiled_layers:
    x = layer(x)

print(x)
```

### Versions

Tried it with both: 2.5.1 and 2.6.0.dev20241119+cu124

cc @chauhang @penguinwu

**Comment by bdhirsh** [comment]
Oh @felipemello1 the recompile is because:

(1) you are compiling two individual layers

(2) on the first layer, your input `x` does not require_grad, and so we don't need to compute gradients for it

(3) on the second layer, the input to that layer is an activation, which we **do** need to compute gradients for.

We need to recompile for the first layer vs the second, because we need to do less compute in the backward of the first layer.

This should cause at most one recompile though - is that a problem for you? Altenratively, you could try compiling both layers together into a single graph to avoid the recompiles
